File: BST.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

""" 
An implementation of a complete binary search tree. The 'complete' part of the 
this tree comes where each node will always have two children, where one will 
either be a value and the other None or both will be None.
"""

from copy import deepcopy


class Leaf(object):

    def __init__(self, value):
        self.value = value
        self.parent = None
        self.left = None
        self.right = None

def option_getter(command):
    def option(tree):
        while True:
            try:
                value = int(raw_input('Enter value > '))
                _leaf = Leaf(value)
                leaf = deepcopy(_leaf)
                return command(tree, leaf) 
            except ValueError:
                print "Enter a number"
    return option

@option_getter
def insert(tree, leaf):
    parent = None
    _tree = tree
    while _tree:
        parent = _tree
        if leaf.value < _tree.value:
            _tree = _tree.left
        else:
            _tree = _tree.right
    leaf.parent = parent
    if parent and leaf.value < parent.value:
        parent.left = leaf
    elif parent and leaf.value >= parent.value:
        parent.right = leaf
    else:
        tree = leaf
    return tree              
 
@option_getter
def delete(tree, leaf):
    if not leaf.left:
       tree = transplant(tree, leaf, leaf.right)
    elif not leaf.right:
       tree = transplant(tree, leaf, leaf.left)
    else:
        _branch = tree_min(leaf.right)
        if not _branch.parent:
            tree = transplant(tree, _branch, _branch.right)
            _branch.right = leaf.right
            _branch.right.parent = _branch
        tree = transplant(tree, leaf, _branch)
        _branch.left = leaf.left
        _branch.left.parent = _branch
    return tree

def transplant(tree, leaf, branch):
    _tree = tree
    if not leaf.parent:
        _tree = branch
    elif leaf == leaf.parent.left:
        leaf.parent.left = branch
    else:
        leaf.parent.right = branch
    if branch:
        branch.parent = leaf.parent
    return tree

@option_getter
def search(tree, leaf):
    if not tree:
        print "Value <%d> not in tree!" % leaf.value
        return
    elif tree.value == leaf.value:
        print "Value <%d> is in tree!" % leaf.value
        return
    while tree:
        if tree.value == leaf.value:
            print "Value <%d> is in tree!" % leaf.value
            return
        elif tree.value < leaf.value:
            tree = tree.right
        else:
            tree = tree.left
    print "Value <%d> is not in tree!" % leaf.value
    return    
    
def tree_walk(tree):
    if tree:
        tree_walk(tree.left)
        print tree.value
        tree_walk(tree.right)

def tree_predecessor(tree):
    if tree.left:
        return tree_max(tree.left)
    parent = tree.parent
    while parent and tree == parent.left:
        tree = parent
        parent = parent.parent
    return parent.value

def tree_successor(tree):
    if tree.right:
        return tree_min(tree.right)
    parent = tree.parent
    while parent and tree == parent.right:
        tree = parent
        parent = parent.parent
    return parent.value

def tree_max(tree):
    while tree.right:
        tree = tree.right
    return tree

def tree_min(tree):
    while tree.left:
        tree = tree.left
    return tree


if __name__ == '__main__':
    tree = None
    while True:
        print "Select option: [1=insert, 2=delete, 3=search, 4=tree_walk, 5=exit]"
        option = raw_input('Enter number of option > ')
        if option == '1':
            tree = insert(tree)
        elif option == '2':
            tree = delete(tree)
        elif option == '3':
            search(tree)
        elif option == '4':
            tree_walk(tree)    
        else:
            break

File: BST.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>


struct leaf {

    struct leaf *parent;
    struct leaf *left;
    struct leaf *right;

    int value;
};

struct leaf *create_leaf(void) {

    struct leaf *fresh_leaf = malloc ( sizeof *fresh_leaf );
    fresh_leaf->parent = NULL;
    fresh_leaf->left = NULL;
    fresh_leaf->right = NULL;
    return fresh_leaf;
}

struct leaf *min_tree(struct leaf **tree) { 

    if ((*tree)->left) {
        min_tree(&(*tree)->left);
    }

    else {
        return *tree; 
    }
}

void transplant(struct leaf **rm_branch, struct leaf **branch) {

    if ((!(*rm_branch)->parent)) {
        (*branch)->parent = NULL;
        (*branch)->left = (*rm_branch)->left;
        (*branch)->left->parent = *branch;
        *rm_branch = *branch; 
    }                         

    else if (*rm_branch == (*rm_branch)->parent->left) {
        (*rm_branch)->parent->left = *branch;
    }

    else {
        (*rm_branch)->parent->right = *branch;
    }

    if (*branch) { 
        (*branch)->parent = (*rm_branch)->parent;  
    }                                             
}

void delete(struct leaf **tree, int value) {

    if (!(*tree)) {
        printf ("Value %d not in tree!\n", value);
    }

    else if ((*tree)->value == value) {

        if ((!(*tree)->left)) {          
            transplant(&(*tree), &(*tree)->right);
        }

        else if ((!(*tree)->right)) {
            transplant(&(*tree), &(*tree)->left); 
        }

        else {

            struct leaf *_branch = malloc ( sizeof *_branch ); 
            _branch = min_tree(&(*tree)->right);

            if (_branch->parent != *tree) { 
                transplant(&_branch, &(_branch)->right);   
                _branch->right = (*tree)->right;          
                _branch->right->parent = _branch;        
            }

            transplant(&(*tree), &_branch); 
            _branch->left = (*tree)->left;     
            _branch->left->parent = _branch;  
        }
    }

    else if ((*tree)->value < value) {
        delete(&(*tree)->right, value);
    }

    else {
        delete(&(*tree)->left, value);
    }
}

void insert(struct leaf **tree, int value) {

    if ((*tree)->value < value) {

        if (!((*tree)->right)) {

            struct leaf *new_leaf = create_leaf();
            new_leaf->parent = *tree;
            new_leaf->value = value;

            (*tree)->right = new_leaf;
        }

        else {
            insert(&(*tree)->right, value);
        }
    }

    else if ((*tree)->value >= value) {

        if (!((*tree)->left)) {

            struct leaf *new_leaf = create_leaf();
            new_leaf->parent = *tree;
            new_leaf->value = value;

            (*tree)->left = new_leaf;
        }

        else {
            insert(&(*tree)->left, value);
        }
    }           
}

void search(struct leaf *tree, int value) {

    if (tree && tree->value != value) {

        if (tree->value < value) {
            search(tree->right, value);
        }

        else {
            search(tree->left, value);
        }
    }

    else if (tree && tree->value == value) {
        printf ("Value %d is in tree!\n", tree->value);
    }

    else {
        printf ("Value %d is not in tree!\n", tree->value);
    }
}

void dump_tree(struct leaf *tree) {

    if (tree) {

        dump_tree(tree->left);
        printf ("Leaf: %d\n", tree->value);
        dump_tree(tree->right);
    }
}

int tree_choice(void) {

    int choice;
    int value;

    struct leaf *tree = NULL;

    while ( true ) {

        printf ("\nPick any of the following choices: [1] insert - [2] delete - [3] search - [4] print tree - [5] exit\n");
        printf ("Enter choice: ");
        scanf ("%d", &choice);
        
        switch (choice) {

            case (1):

                printf ("Enter value: ");
                scanf ("%d", &value);

                if (!(tree)) {
                    tree = create_leaf(); 
                    tree->value = value;
                }

                else {
                    insert(&tree, value);
                }

                break;
            
            case (2):

                printf ("Enter value: ");
                scanf ("%d", &value);
                delete(&tree, value);

                break;

            case (3):

                printf ("Enter value: ");
                scanf ("%d", &value);
                search(tree, value);

                break;

            case (4):

                dump_tree(tree); 

                break;

            case (5):
                return 0;
        } 
    }                 
}

int main(void) {

    printf ("\n -*- Binary Search Tree -*- \n");
    tree_choice();

    return 0;
}
